src/server/terrain/chunkgen/StalagmiteGenerator.zig

```diff
@@ -0,0 +1,139 @@
+const std = @import("std");
+
+const main = @import("root");
+const random = main.random;
+const JsonElement = main.JsonElement;
+const terrain = main.server.terrain;
+const CaveMap = terrain.CaveMap;
+const CaveBiomeMap = terrain.CaveBiomeMap;
+const vec = main.vec;
+
+pub const id = "cubyz:stalagmite";
+
+pub const priority = 16384;
+
+pub const generatorSeed = 0x2ba5bef20d6153a5;
+
+var stalagmite: u16 = 0;
+
+const surfaceDist = 2; // How far away stalagmite can spawn from the wall.
+
+pub fn init(parameters: JsonElement) void {
+	_ = parameters;
+	
+	stalagmite = main.blocks.getByID("cubyz:limestone");
+}
+
+pub fn deinit() void {
+
+}
+
+pub fn generate(worldSeed: u64, chunk: *main.chunk.ServerChunk, caveMap: CaveMap.CaveMapView, biomeMap: CaveBiomeMap.CaveBiomeMapView) void {
+	if(chunk.super.pos.voxelSize > 2) return;
+	const size = chunk.super.width;
+	// Generate caves from all nearby chunks:
+	var x = chunk.super.pos.wx -% main.chunk.chunkSize;
+	while(x != chunk.super.pos.wx +% size +% main.chunk.chunkSize) : (x +%= main.chunk.chunkSize) {
+		var y = chunk.super.pos.wy -% main.chunk.chunkSize;
+		while(y != chunk.super.pos.wy +% size +% main.chunk.chunkSize) : (y +%= main.chunk.chunkSize) {
+			var z = chunk.super.pos.wz -% main.chunk.chunkSize;
+			while(z != chunk.super.pos.wz +% size +% main.chunk.chunkSize) : (z +%= main.chunk.chunkSize) {
+				var seed = random.initSeed3D(worldSeed, .{x, y, z});
+				considerCoordinates(x, y, z, chunk, caveMap, biomeMap, &seed);
+			}
+		}
+	}
+}
+
+fn distSqr(x: f32, y: f32, z: f32) f32 {
+	return x*x + y*y + z*z;
+}
+
+fn considerStalagmite(x: i32, y: i32, z: i32, chunk: *main.chunk.ServerChunk, caveMap: CaveMap.CaveMapView, seed: *u64) void {
+	const relX: f32 = @floatFromInt(x -% chunk.super.pos.wx);
+	const relY: f32 = @floatFromInt(y -% chunk.super.pos.wy);
+	const relZ: f32 = @floatFromInt(z -% chunk.super.pos.wz);
```

Comment:
I think it would make sense to give them a random offset from the base block, to avoid having all of them look perfectly symmetrical.